java/src/main/java/io/cucumber/java/JavaHookDefinition.java

```diff
@@ -22,41 +29,88 @@
+        if (null == annotations || annotations.length == 0) {
+            return new Class<?>[0];
+        }
+        Class<?>[] relevant = null;
+        for (Annotation annotation : annotations) {
+            Class<? extends Annotation> annotationType = annotation.annotationType();
+            if (Before.class.equals(annotationType) || After.class.equals(annotationType)) {
+                return new Class<?>[] {Before.class, After.class};
+            }
+            if (BeforeStep.class.equals(annotationType) || AfterStep.class.equals(annotationType)) {
+                relevant = new Class<?>[] {BeforeStep.class, AfterStep.class};
             }
         }
+        return relevant;
+    }
 
-        return method;
+    private static Set<Class<?>> getAcceptedTypes(Class<?> annotationType) {
+        if (null == annotationType) {
+            return Collections.emptySet();
+        }
+        Set<Class<?>> accepted = new HashSet<>();
+        if (Before.class.equals(annotationType) || After.class.equals(annotationType)) {
+            accepted.add(Scenario.class);
+        }
+        if (BeforeStep.class.equals(annotationType) || AfterStep.class.equals(annotationType)) {
+            accepted.addAll(Arrays.asList(Scenario.class, Step.class));
+        }
+        return accepted;
     }
 
-    private static InvalidMethodSignatureException createInvalidSignatureException(Method method) {
-        return builder(method)
-                .addAnnotation(Before.class)
-                .addAnnotation(After.class)
-                .addAnnotation(BeforeStep.class)
-                .addAnnotation(AfterStep.class)
-                .addSignature("public void before_or_after(io.cucumber.java.Scenario scenario)")
-                .addSignature("public void before_or_after()")
-                .build();
+    private static void checkParameter(Method method, Class<?> parameterType, Class<?>[] annotationTypes) {
+        for (Class<?> annotationType : annotationTypes) {
+            Set<Class<?>> acceptedTypes = getAcceptedTypes(annotationType);
+            if (!(Object.class.equals(parameterType) || acceptedTypes.contains(parameterType))) {
+                throw createInvalidSignatureException(method, annotationTypes, acceptedTypes);
+            }
+        }
+    }
+
+    private static InvalidMethodSignatureException createInvalidSignatureException(Method method,
+                                                                                   Class<?>[] methodAnnotations,
+                                                                                   Set<Class<?>> acceptedTypes) {
+        InvalidMethodSignatureException.InvalidMethodSignatureExceptionBuilder builder = builder(method);
+        if (methodAnnotations.length == 0) {
+            throw new IllegalArgumentException(
+                    "Method should be annotated with one of: @Before, @After, @BeforeStep, @AfterStep");
```

Comment:
This can't happen. If a method is not annotated there will be no instance of `JavaHookDefinition` created.